### ISSUE TITLE
Missing elevation in AWS_station_locations.csv

### ISSUE THREAD

**Issue by BaptisteVandecrux** [GEUS-Glaciology-and-Climate/PROMICE-AWS-data-issues#46]
https://geusgitlab.geus.dk/glaciology-and-climate/promice/aws-l3/-/blob/main/AWS_station_locations.csv

- [ ] KAN_M
- [ ] SCO_L
- [ ] SCO_U
- [ ] TAS_A

**Comment by BaptisteVandecrux** [comment]
These four stations have a disfunctional GPS. The lat/lon currently used in [AWS_station_locations.csv](https://geusgitlab.geus.dk/glaciology-and-climate/promice/aws-l3/-/blob/main/AWS_station_locations.csv) comes from the Iridium emails. Starting some work here: https://github.com/GEUS-Glaciology-and-Climate/pypromice/tree/fixing-missing-elevation

**Comment by BaptisteVandecrux** [comment]
**The problem:**

When GetBUFR was being called, a linear fit based on the last three months of lat, lon and alt was calculated. For latitude and longitude, coordinates contained in the Iridium emails were used to gap-fill missing observation of latitude and longitude, meaning that there was always data to fit a linear function on. For altitude, however, a failing GPS over more than 3 months would  lead to the absence of observed altitude to build a linear function. In those cases, altitude is left as NaN in the BUFR file sent to DMI/WMO and saved in [AWS_station_locations.csv](https://geusgitlab.geus.dk/glaciology-and-climate/promice/aws-l3/-/blob/main/AWS_station_locations.csv).

**The solution:**
My suggestion is to make a first attempt to interpolate lat, lon and altitude from the last three months and if altitude could not be estimated from that first try, we extrapolated the altitude for all the timestamps where it's missing, based on the entire record of measured altitude (that could be from several years back).

The pros:
This will always provide an updated, best estimation of the station altitude.

The cons:
Some of the stations have a failing GPS for many years (e.g. SCO_U since 2012). We therefore estimate the position in 2023 based on 2008-2012 data. I believe it is acceptable because the stations' downward motion is rather steady.

Some questions for for @patrickjwright :
- Are the BUFR files with missing altitude sent to DMI?
- If yes, would DMI prefer NaNs or extrapolated elevations?
- (Very minor) The logic of the linear interpolation was hard to follow: A linear model is fitted, but then it is used to estimate lat, lon or alt only for the timestamps where that variable was actually observed. Then the most recent index where all three lat, lon and lat are available is used as a the position's timestamp. Why not using that linear function for all timestamps (also the ones when lat, lon or alt observation might be missing), up to the one that needs to be transmitted to DMI?

Here is the extrapolation result for the four problematic stations:
![image](https://user-images.githubusercontent.com/35140661/236680193-aad39828-1a5b-4403-89cb-2b0b3c96e09e.png)
![image](https://user-images.githubusercontent.com/35140661/236680203-884b8397-274d-41d9-b1cc-d990c0d166a1.png)
![image](https://user-images.githubusercontent.com/35140661/236680216-f22337e2-2032-4813-be0c-72cfff6b2695.png)
![image](https://user-images.githubusercontent.com/35140661/236680226-8a3890f5-2729-48ad-9fd6-2e95671d53a3.png)

**Comment by patrickjwright** [comment]
@BaptisteVandecrux Overall I think it is fine to extrapolate elevation from the last known point. Glad you are looking into it! It would also be good to figure out why elevation is not included in the Iridium email text, and get that included.

To answer your questions:

_Are the BUFR files with missing altitude sent to DMI?_

No. You can see [here](https://github.com/GEUS-Glaciology-and-Climate/pypromice/blob/c3a934f309e427e7d3fb667b01d6b0bf6122e6bd/src/pypromice/postprocess/csv2bufr.py#L489) in `csv2bufr.min_data_check` that we currently require all three of lat, lon and elev to submit to DMI. I opted to be strict with this, since DMI indicated that station metadata (especially elev) would be important.

_If yes, would DMI prefer NaNs or extrapolated elevations?_

I think DMI would be fine with extrapoloated elevation, since it is better than nothing. The thing they want to avoid is elevation that jumps up and down from transmission to transmission (thus the linear-fit smoothing).

_(Very minor) The logic of the linear interpolation was hard to follow: A linear model is fitted, but then it is used to estimate lat, lon or alt only for the timestamps where that variable was actually observed. Then the most recent index where all three lat, lon and lat are available is used as a the position's timestamp. Why not using that linear function for all timestamps (also the ones when lat, lon or alt observation might be missing), up to the one that needs to be transmitted to DMI?_

My goal with the linear regression was not to create any new interpolated data, but only to reduce noise in the transmitted data. I wanted to submit the actual measured timestamp, but use the linear regression to make sure each submitted observation followed a smooth trajectory. If an observation was not available in the previous 2 days (or if we had other variables, but no position data), then we do not submit to DMI. Creating interpolated data using the regression model was a step I had not yet taken.

**Also, an important detail**: Your problem statement is not quite accurate:

_When GetBUFR was being called, a linear fit based on the last three months of lat, lon and alt was calculated. For latitude and longitude, coordinates contained in the Iridium emails were used to gap-fill missing observation of latitude and longitude, meaning that there was always data to fit a linear function on. For altitude, however, a failing GPS over more than 3 months would lead to the absence of observed altitude to build a linear function. In those cases, altitude is left as NaN in the BUFR file sent to DMI/WMO and saved in [AWS_station_locations.csv](https://geusgitlab.geus.dk/glaciology-and-climate/promice/aws-l3/-/blob/main/AWS_station_locations.csv)._

Yes, the Iridium text positions are used to gap-fill using the `combine_first()` function. However, the 3 months of data for linear regression is previous to the last received transmission (could be months or years ago). If we received a transmission in the last two days, we run `csv2bufr.linear_fit` directly and submit to DMI. But if not, then we run `csv2bufr.fetch_old_positions` only for the purpose of writing the last known position to `AWS_station_locations.csv` (see [here](url)). The stations with missing elevation should mean that the GPS elevation is missing entirely.... in which case, even interpolation should not provide data.

We also run `fetch_old_positions` in the case that a station was skipped due to a setting in `wmo_config.py` (see [here](https://github.com/GEUS-Glaciology-and-Climate/pypromice/blob/c3a934f309e427e7d3fb667b01d6b0bf6122e6bd/bin/getBUFR#L273)). In summary, we only submit to DMI if we have a recent transmission, but we will attempt to get position information for every station for writing to the csv file (whether skipped, or whether the data is old).

With what you are proposing, the changes would be:
1. The position csv file would have an estimated position based on a linear trend, rather than the last known position.
2. We would actually submit to DMI if we have other variables (but missing position) by extending trends from the last valid position data.

**Comment by patrickjwright** [comment]
To quote myself:
> But if not, then we run csv2bufr.fetch_old_positions only for the purpose of writing the last known position to AWS_station_locations.csv (see [here](https://github.com/GEUS-Glaciology-and-Climate/PROMICE-AWS-data-issues/issues/url)). The stations with missing elevation should mean that the GPS elevation is missing entirely.... in which case, even interpolation should not provide data.

Maybe this is not true.... if the last transmission was months ago, but the GPS altitude was missing from many months before that.... then we could end up without GPS altitude when we take the previous 90 days. Even though it is present somewhere in the historical record......?

**Comment by BaptisteVandecrux** [comment]
Thanks for the explanation!

> Maybe this is not true.... if the last transmission was months ago, but the GPS altitude was missing from many months before that.... then we could end up without GPS altitude when we take the previous 90 days. Even though it is present somewhere in the historical record......?

Yes exactly!

The four stations named above are still transmitting good data. It is only their GPS that is not working.
So for those four stations we have potentially temperature, pressure, latitude, longitude (from the iridium email) but not the elevation. Now I realize that this situation is not handled is 'fetch_old_positions' but directly in 'GetBUFR':

In the lines below, the code first writes the position smoothed from the last three months of data. If the station's GPS is down, then lat, lon is recovered from the email but elevation is still missing. Then in our case, the 'min_data_check' finds out that elevation is missing then it just adds to the list of failed to process, and continues without trying to fill the missing coordinates.

https://github.com/GEUS-Glaciology-and-Climate/pypromice/blob/c3a934f309e427e7d3fb667b01d6b0bf6122e6bd/bin/getBUFR#L234-L249

I'm therefore suggesting to add a 'fetch_old_position' there as well (link below from the PR branch):

https://github.com/GEUS-Glaciology-and-Climate/pypromice/blob/c4c9a069fbaea5339f8ecb824a1ee32ecc0b71ce/bin/getBUFR#L240-L247

With this fix, we continue to provide the same file quality to DMI, but the AWS_positions.csv should now receive elevation coordinates for all stations. We can then decide later whether we want to provide DMI altitudes extrapolated over many years.

**Comment by BaptisteVandecrux** [comment]
Now coordinates for the stations will always be available in [AWS_metadata.csv](https://geusgitlab.geus.dk/glaciology-and-climate/promice/aws-l3/-/blob/main/AWS_metadata.csv).

[AWS_station_locations.csv](https://geusgitlab.geus.dk/glaciology-and-climate/promice/aws-l3/-/blob/main/AWS_station_locations.csv). was renamed to [AWS_latest_locations.csv](https://geusgitlab.geus.dk/glaciology-and-climate/promice/aws-l3/-/blob/main/AWS_latest_locations.csv) and has the limitation of being built from transmissions only. So, for some stations it cannot estimate the current coordinates. 

In the future, we could update AWS_metadata.csv using the entire historical record so we are sure we provide latest coordinates for all stations. Then we won't need AWS_latest_locations.csv anymore.
This further code improvement will be handled on https://github.com/GEUS-Glaciology-and-Climate/pypromice/issues/142

In the meantime, the all coordinates will remain available in AWS_metadata.csv on the dataverse.
